components/equihash/tromp/equi_miner.c

```diff
@@ -228,42 +245,58 @@ struct equi {
   bsizes *nslots; // PUT IN BUCKET STRUCT
   proof *sols;
   au32 nsols;
-  u32 nthreads;
   u32 xfull;
   u32 hfull;
   u32 bfull;
-  pthread_barrier_t barry;
 };
 typedef struct equi equi;
+  void equi_clearslots(equi *eq);
   equi *equi_new(
-    const u32 n_threads,
     blake2b_clone blake2b_clone,
     blake2b_free blake2b_free,
     blake2b_update blake2b_update,
     blake2b_finalize blake2b_finalize
   ) {
     assert(sizeof(hashunit) == 4);
     equi *eq = malloc(sizeof(equi));
-    eq->nthreads = n_threads;
     eq->blake2b_clone = blake2b_clone;
     eq->blake2b_free = blake2b_free;
     eq->blake2b_update = blake2b_update;
     eq->blake2b_finalize = blake2b_finalize;
-    const int err = pthread_barrier_init(&eq->barry, NULL, eq->nthreads);
-    assert(!err);
+
     alloctrees(&eq->hta);
     eq->nslots = (bsizes *)htalloc_alloc(&eq->hta, 2 * NBUCKETS, sizeof(au32));
     eq->sols   =  (proof *)htalloc_alloc(&eq->hta, MAXSOLS, sizeof(proof));
+
+    // C malloc() does not guarantee zero-initialized memory (but calloc() does)
+    eq->blake_ctx = NULL;
+    eq->nsols = 0;
+    equi_clearslots(eq);
+
     return eq;
   }
   void equi_free(equi *eq) {
+    if (eq == NULL) {
+      return;
+    }
+
     dealloctrees(&eq->hta);
+
     free(eq->nslots);
     free(eq->sols);
     eq->blake2b_free(eq->blake_ctx);
+    // Avoid use-after-free and double-free
+    eq->nslots = NULL;
+    eq->sols = NULL;
+    eq->blake_ctx = NULL;
+
     free(eq);
   }
   void equi_setstate(equi *eq, const BLAKE2bState *ctx) {
+    if (eq->blake_ctx) {
+      eq->blake2b_free(eq->blake_ctx);
+    }
+
     eq->blake_ctx = eq->blake2b_clone(ctx);
     memset(eq->nslots, 0, NBUCKETS * sizeof(au32)); // only nslots[0] needs zeroing
     eq->nsols = 0;
```

Comment:
This allows us to re-use an `eq` instance, and we've just cleared the slot memory.

```suggestion
    equi_clearslots(eq);
    eq->nsols = 0;
```